### PR TITLE
[Script] Factorize out Generic Free Monad Runner

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**

### CODE REVIEW

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala [hunk 1]
```diff
@@ -33,31 +67,48 @@ private[lf] class Runner(
   private val machine =
     Speedy.Machine.fromPureSExpr(
       unversionedRunner.extendedCompiledPackages,
-      unversionedRunner.script.expr,
+      SEValue(SValue.SUnit),
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala:70]: prefer to set up the machine just before starting interpretation.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala [hunk 2]
```diff
@@ -33,31 +67,48 @@ private[lf] class Runner(
...
       iterationsBetweenInterruptions = 100000,
       traceLog = traceLog,
       warningLog = warningLog,
     )(Script.DummyLoggingContext)
 
-  @scala.annotation.tailrec
-  final def stepToValue(): Either[RuntimeException, SValue] = {
-    val result = machine.run()
+  def consume[X](result: => Result[X])(implicit
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala:76]: Convert Result to Future

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala [hunk 3]
```diff
@@ -33,31 +67,48 @@ private[lf] class Runner(
...
+              case Success(x) => consume(resume(Right(x)))
+              case Failure(err: RuntimeException) => consume(resume(Left(err)))
+              case Failure(err) => Future.failed(err)
+            }
         }
     }
   }
 
+  @scala.annotation.nowarn("msg=dead code following this construct")
+  final def runExpr(expr: SExpr): Result[SValue] = {
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala:99]: convert SResult to Result

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala [hunk 4]
```diff
@@ -15,13 +15,47 @@ import com.daml.lf.engine.script.ledgerinteraction.{
...
+
+sealed abstract class Result[+X] extends Product with Serializable {
+  def transform[Y](f: Either[RuntimeException, X] => Result[Y]): Result[Y]
+  def flatMap[Y](f: X => Result[Y]): Result[Y] = transform(_.fold(Result.Error, f))
+  def map[Y](f: X => Y): Result[Y] = flatMap(x => Result.Done(f(x)))
+}
+
+object Result {
+
+  def fromEither[X](either: Either[String, X]): Result[X] =
```
Comment (samuel-williams-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Runner.scala:32]: Would probably make this clear that its a conversion failure in the name

**Review by samuel-williams-da — COMMENTED** [review]
I like this transformation so far - you've pushed all the computation up to `consume`.
I assume the goal is, in some way, to modify the ledger client interface such that every method now takes the continuation from the question directly, rather than calling to the current `runCmdQuestion`? This'll require refactoring ScriptF as well to push the continuations through manually? That, or refactoring such that it provides a "send" and "receive" that handles the translation logic it needs before and after the call to the ledger client.
My previous script changes where to keep boiler plate low - and passing around continuations everywhere was part of that. If you can find a nice way to push control to the ledger clients without having explicit continuations everywhere, that'd be neat

**Comment by samuel-williams-da** [comment]
Another option would be to have `consume` be the stopping point for a generic monad you pass down from the ledger clients, such that IDE runs in Identity, and the others run in whatever monad they need, this makes the continuation passing implicit in a way that is familiar.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/Free.scala [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    def apply[Q](q: Q, continue: SValue): Question[SExpr, Q] =
+      new Question(q, continue, Final(_))
+  }
+}
+
+case class ConversionError(message: String) extends RuntimeException(message)
+final case class InterpretationError(error: SError.SError)
+    extends RuntimeException(s"${Pretty.prettyError(error).render(80)}")
+
+object Free {
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/Free.scala:81]: This could be use almost as such by (a new version of ?) triggers

**Comment by remyhaemmerle-da** [comment]
@carlpulley-da @samuel-williams-da. Here is a total new version, could you have a look ?

**Review by samuel-williams-da — APPROVED** [review]
This is very clean - I like it! I'd guess we'll refactor triggers down the line to use this? (if triggers remains in daml)

I almost also consider a slight change to interruption - someone using the runner doesn't need to know that interruption is a state that it can end in. It might be reasonable to have the user pass in handler when calling Free, so they can think of the Free Result they're parsing as "identical" to the Free they already know (from generic free monads or the implementation in daml)
It depends on if we ever expect interruption resuming to be based on the current expression, which i doubt we would, given the lack of control over when interruptions happen

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/Free.scala [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+      copy(resume = resume(_).transform(f))
+
+    override def remapQ[R](f: Q => Result[ErrOr[SExpr], R]): Result[X, R] =
+      f(q).flatMap(resume(_).remapQ(f))
+  }
+
+  val Unit = Result.Final(Right(()))
+}
+
+case class ConversionError(message: String) extends RuntimeException(message)
```
Comment (carlpulley-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/Free.scala:65]: Surely, we should have this as final and private? Particularly worried that outside users might try and compare multiple instances on this as we are in an abstract class.

**Comment by remyhaemmerle-da** [comment]
> someone using the runner doesn't need to know that interruption is a state that it can end in

I add a method in Result 
```scala
def run[R >: Q, B <: A](
        answer: R => Result[B, R, B],
        canceled: () => Option[RuntimeException],
    ): ErrOr[X] = {
```

**Comment by samuel-williams-da** [comment]
> > someone using the runner doesn't need to know that interruption is a state that it can end in
> 
> I add a method in Result
> 
> ```scala
> def run[R >: Q, B <: A](
>         answer: R => Result[B, R, B],
>         canceled: () => Option[RuntimeException],
>     ): ErrOr[X] = {
> ```

Is the expectation that `answer` would be a blocking call?

**Pull request merged**